services/graph/pkg/service/v0/driveitems.go

```diff
@@ -691,15 +684,29 @@ func (g Graph) DeletePermission(w http.ResponseWriter, r *http.Request) {
 	}
 
 	ctx := r.Context()
-	isUserPermission := true
+	var permissionType PermissionType
 
-	// Check if the id is referring to a User Share
-	sharedResourceID, err := g.getUserPermissionResourceID(ctx, permissionID)
-	var errcode errorcode.Error
-	if err != nil && errors.As(err, &errcode) && errcode.GetCode() == errorcode.ItemNotFound {
-		// there is no user share with that ID, so lets check if it is referring to a public link
-		isUserPermission = false
-		sharedResourceID, err = g.getLinkPermissionResourceID(ctx, permissionID)
+	// Check if the ID is referring to a public share
+	sharedResourceID, err := g.getLinkPermissionResourceID(ctx, permissionID)
+	if err == nil {
+		permissionType = Public
+	}
+
+	// If the item id is referring to a space root an this is not a public share
+	// we have to deal with space permissions
+	if permissionType == Unknown && IsSpaceRoot(&itemID) {
+		permissionType = Space
+		sharedResourceID = &itemID
+		err = nil
+	}
+
+	// If this is neither a public share nor a space permission, check if this is a
+	// user share
+	if permissionType == Unknown {
+		sharedResourceID, err = g.getUserPermissionResourceID(ctx, permissionID)
+		if err == nil {
+			permissionType = User
+		}
```

Comment:
```suggestion
	sharedResourceID, err := g.getLinkPermissionResourceID(ctx, permissionID)
	switch {
	// Check if the ID is referring to a public share
	case err == nil:
		permissionType = Public

	// If the item id is referring to a space root an this is not a public share
	// we have to deal with space permissions
	case IsSpaceRoot(&itemID):
		permissionType = Space
		sharedResourceID = &itemID

	// If this is neither a public share nor a space permission, check if this is a
	// user share
	default:
		sharedResourceID, err = g.getUserPermissionResourceID(ctx, permissionID)
		if err == nil {
			permissionType = User
		}
	}
```

Wouldn't this be simpler with a switch case? Or am I misunderstanding the logic? :smile:

Reply:
I guess you're right. Seems I stared at this code for too long.